EIP-2831 Transaction Replacement Message Type (#2831)
ethereum/EIPs: EIPS/eip-2831.md

---
eip: 2831
title: Transaction Replacement Message Type
author: Gregory Markou (@GregTheGreek)

>  discussions-to: https://ethereum-magicians.org/t/eip-2831-transaction-replacement-message-type/4448

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-2831-transaction-replacement-message-type/4448
```

> discussions-to: [eth magicians](https://ethereum-magicians.org/t/eip-2831-transaction-replacement-message-type/4448)

Reviewer 2: EIPs normally just have a raw link here.  The final EIP renderer may try to pretty this at some point in the future, so I think it is good to keep it as a raw link everywhere.
```suggestion
 discussions-to: https://ethereum-magicians.org/t/eip-2831-transaction-replacement-message-type/4448
```

status: Draft
type: Standards Track
category: Interface
created: 2020-07-26
requires: 1193
---

## Summary

> An extension to the JavaScript Ethereum Proider API ([EIP-1193](https://eips.ethereum.org/EIPS/eip-1193)) this creates a new message type in the event a transaction replacement occurs.

Reviewer 2: ```suggestion
An extension to the JavaScript Ethereum Proider API ([EIP-1193](./eip-1193.md)) this creates a new message type in the event a transaction replacement occurs.
```

## Abstract

Exert from EIP-1193
> A common convention in the Ethereum web application ("dapp") ecosystem is for key management software ("wallets") to expose their API via a JavaScript object in the web page.
This object is called "the Provider".

> Many ingenious developments have been made by wallet developers to improve the overall user experience while interacting with the Ethereum blockchain. One specific innovation was the tx replacement, offering users the ability to effectively cancel a previously sent transaction.

Reviewer 2: ```suggestion
Many ingenious developments have been made by wallet developers to improve the overall user experience while interacting with the Ethereum blockchain. One specific innovation was transaction replacement, offering users the ability to effectively cancel a previously sent transaction.
```

> Transaction replacement is not a new concept, but unfortunately causes major user experience problems for dapp developers as the replaced transaction is near impossible to track. 

Reviewer 2: This all feels like it should be in the `## Motivation` section, which appears to be missing.

This EIP formalizes a way for both providers and dapp developers to track transaction replacements seamlessly.

> This specification does not cover how one would utilize the transaction replacement message.

Reviewer 2: I recommend adding to the abstract a very brief code description of the event being added, like the TypeScript signature

> This specification does not cover how one would utilize the transaction replacement message.

Reviewer 2: Consider removing the statement that says what the EIP doesn't do.   The EIP also doesn't make me coffee.  😉

Author: LOOL Love ok :)

I took eip 1193 as a template so it may be outdated for that reason

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).

> Comments like this are non-normative.

### Definitions

_This section is non-normative._

- Provider
  - A JavaScript object made available to a consumer, that provides access to Ethereum by means of a Client.
- Client
  - An endpoint that receives Remote Procedure Call (RPC) requests from the Provider, and returns their results.
- Wallet
  - An end-user application that manages private keys, performs signing operations, and acts as a middleware between the Provider and the Client.
- Remote Procedure Call (RPC)

>   - A Remote Procedure Call (RPC), is any request submitted to a Provider for some procedure that is to be processed by a Provider, its Wallet, or its Client.

Reviewer 2: These definitions are not present in the rest of the specification, so they probably shouldn't be included in the EIP.

- Transaction Replacement

>   - Submitting a transaction with both: the same nonce and a 10% increase in the gas price, of a previous transaction which a user no longer wishes to send. This must occur before the original transaction is included in the blockchain.

Reviewer 2: This section header says it is non-normative, but this definitely seems like something that should be normative.

### Interface

The `data` parameter contains two values: `oldTx` and `newTx`,the transaction hash that's being superseded and the transaction hash of the successor, respectively. 

```JavaScript
interface TxReplacement extends ProviderMessage {
  readonly type: 'tx_replacement';
  readonly data: {
    readonly oldTx: string;

>     readonly newTx: string;

Reviewer 2: Recommend defining more clearly how these are encoded.  Is it a number as a string, a hex number as a string, is 0x prefixing mandatory, optional, or disallowed?

  };
}
```

> ## Rationale

Reviewer 2: Recommend explaining why ProviderMessage was used rather than a new event type. A new event type seems much more appropriate here than throwing it into the generic ProviderMethod (which IMO is intended for random non-standard stuff).

> The purpose of `tx_replacement` is to provide a user experience improvement to both wallet users and dapp developers. It can be implemented without breaking changes and leverages the existing framework set out by EIP-1193.

Reviewer 2: This statement feels like a repeat of the motivation/abstract.  The purpose of the Rationale section is (usually) to describe why you made certain decisions within the specification, like why you decided to use `ProviderMessage` instead of a new message type.

> ## Backwards Compatibility

Reviewer 2: Recommend explaining why `ProviderMessage` was used rather than a new event type.  A new event type seems much more appropriate here than throwing it into the generic `ProviderMethod` (which IMO is intended for random non-standard stuff).
